File: dataDevops/ambari-versions/ambari-3.0.0.0-SNAPSHOT/ambari-web/app/utils/file_utils.js
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

module.exports = {

  fileTypeMap: {
    csv: 'text/csv',
    json: 'application/json'
  },

  /**
   * download text file
   * @param data {String}
   * @param fileType {String}
   * @param fileName {String}
   */
  downloadTextFile: function (data, fileType, fileName) {
    if ($.browser.msie && $.browser.version < 10) {
      this.openInfoInNewTab(data);
    } else if (typeof safari !== 'undefined') {
      this.safariDownload(data, fileType, fileName);
    } else {
      try {
        var blob = new Blob([data], {
          type: (this.fileTypeMap[fileType] || 'text/' + fileType) + ';charset=utf-8;'
        });
        saveAs(blob, fileName);
      } catch (e) {
        this.openInfoInNewTab(data);
      }
    }
  },

  /**
   * download multiple files archived in ZIP
   * @param {object[]} files
   */
  downloadFilesInZip: function(files) {
    const zip = new JSZip();
    for (const file of files) {
      const blob = new Blob([file.data], {
        type: (this.fileTypeMap[file.type] || 'text/' + file.type) + ';charset=utf-8;'
      });
      zip.file(file.name, blob);
    }
    zip.generateAsync({type:"blob"})
      .then(function(content) {
        saveAs(content, "blueprint.zip");
      });
  },

  /**
   * open content of text file in new window
   * @param data {String}
   */
  openInfoInNewTab: function (data) {
    var newWindow = window.open('');
    var newDocument = newWindow.document;
    newDocument.write(data);
    newWindow.focus();
  },

  /**
   * Hack to dowload text data in Safari
   * @param data {String}
   * @param fileType {String}
   * @param fileName {String}
   */
  safariDownload: function (data, fileType, fileName) {
    var file = 'data:attachment/' + fileType + ';charset=utf-8,' + encodeURI(data);
    var linkEl = document.createElement("a");
    linkEl.href = file;
    linkEl.download = fileName;

    document.body.appendChild(linkEl);
    linkEl.click();
    document.body.removeChild(linkEl);
  },

  fileNameFromPath: function(path) {
    return path.split('/').slice(-1);
  }

};

File: market/tis-versions/tis-3.6.0/tis/tis-common/src/main/java/com/qlangtech/tis/pubhook/common/FileUtils.java
/**
 *   Licensed to the Apache Software Foundation (ASF) under one
 *   or more contributor license agreements.  See the NOTICE file
 *   distributed with this work for additional information
 *   regarding copyright ownership.  The ASF licenses this file
 *   to you under the Apache License, Version 2.0 (the
 *   "License"); you may not use this file except in compliance
 *   with the License.  You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 */
package com.qlangtech.tis.pubhook.common;

import com.qlangtech.tis.manage.common.TisUTF8;
import java.io.File;
import java.io.IOException;
import java.io.RandomAccessFile;

/**
 * @author 百岁（baisui@qlangtech.com）
 * @date 2012-2-15
 */
public class FileUtils {

//    public static String readLastLine(File file) {
//        RandomAccessFile randomAccess = null;
//        try {
//            randomAccess = new RandomAccessFile(file, "r");
//            boolean eol = false;
//            int c = -1;
//            long fileLength = randomAccess.length();
//            long size = 1;
//            ww: while (!eol) {
//                long offset = fileLength - (size++);
//                randomAccess.seek(offset);
//                switch(c = randomAccess.read()) {
//                    case -1:
//                    case '\n':
//                    case '\r':
//                        randomAccess.seek(offset + 1);
//                        break ww;
//                }
//            }
//            return randomAccess.readLine();
//        } catch (IOException e) {
//            throw new RuntimeException(e);
//        } finally {
//            try {
//                randomAccess.close();
//            } catch (IOException e) {
//            }
//        }
//    }

    /**
     * 在文件最后追加一行记录
     *
     * @param file
     * @param line
     */
    public static void append(File file, String line) {
        try {
            org.apache.commons.io.FileUtils.write(file, ("\r\n" + line), TisUTF8.get(), true);
        } catch (IOException e) {
            throw new RuntimeException(file.getAbsolutePath(), e);
        }
    // OutputStream wirter = null;
    // try {
    //
    // wirter = new FileOutputStream(file, true);
    // wirter.write(("\r\n" + line).getBytes());
    // } catch (Exception e) {
    // throw new RuntimeException(e);
    // } finally {
    // try {
    // wirter.close();
    // } catch (IOException e) {
    // }
    // }
    }

//    public static void main(String[] arg) {
//        append(new File("D:\\tmp\\test.txt"), "11");
//    }
}
